{% extends 'gendep/base.html' %}
{% load staticfiles %}
{% block content %}   
<h1>Frequently Asked Questions</h1>

<ul>
<li><a href="#effectsize">How should I interpret the effect size?</a></li>
<li><a href="#boxplots">How should I interpret the box plots?</a></li>
<li><a href="#dependencies">What dependencies are included?</a></li>
<li><a href="#gene_altered">When is a driver gene considered 'altered'?</a></li>
<li><a href="#genotyping">Where does the genotyping data come from?</a></li>
<li><a href="#json">How can I access this data from scripts, as JSON web service?</a></li>
<li><a href="#downloaddb">How can I download the whole CancerGD database to interrogate locally?</a></li>
</ul>


<a name="effectsize"></a><h2>How should I interpret the effect size?</h2>

<p>The effect size presented in the database is the <i>Common Language Effect Size</i>, equivalent to the <i>Probability of Superiority</i> and the <i>Area Under the ROC Curve</i>. It indicates the probability that a cell line with an alteration in a particular <a href="{% url 'gendep:drivers' %}">driver gene</a> will be more sensitive to a given RNAi reagent than a cell line without that alteration. For instance, the effect size in Campbell <i>et al</i> for the dependency of <i>ERBB2</i> amplified cell lines upon <i>MAP2K3</i> siRNA is 81%. This indicates that if two cell lines are chosen at random, one with <i>ERBB2</i> amplification and one without, 83% of the time the cell line with the <i>ERBB2</i> amplification will be more sensitive to the <i>MAP2K3</i> siRNA.</p>

<p>Note that the <i>Common Language Effect Size</i> does not indicate the magnitude of the difference in sensitivity between the altered and wild-type group of cell lines. It merely indicates that one group is more sensitive than the other, not how much more sensitive.</p>

<a name="dependencies"><h2>What dependencies are included?</h2></a>

<p>We include all dependencies that have a <i>nominally significant</i> p-value less than 0.05 and a <i>Common Language Effect Size</i> of greater than 65%. By <i>nominally significant</i> we mean that the p-value has not been corrected for multiple hypothesis testing. We, and others, have found that analysing all of the nominally significant dependencies associated with a <a href="{% url 'gendep:drivers' %}">driver gene</a> can provide more insight than analysing only those that survive correction for multiple hypothesis testing.</p>

<a name="boxplots"><h2>How should I interpret the box plots?</h2></a>

<center><img src="{% static 'gendep/images/FAQ_boxplot.png' %}" alt="Box plot"/></center>

<p>Each <a href="https://en.wikipedia.org/wiki/Box_plot" title="Box plot">box plot</a> shows the sensitivity of cell lines partitioned according to the status of a particular <a href="{% url 'gendep:drivers' %}">driver gene</a> to RNAi reagents targeting a particular gene. For instance above we show the box plot indicating a dependency between <i>ERBB2</i> amplification and sensitivity to <i>MAP2K3</i> siRNA. The cell lines featuring the alteration in question are displayed on the right and the cell lines without the alteration are on the left.  Each coloured circle represents a cell line and the position along the y-axis indicates how sensitive that cell line is to the RNAi reagents targeting the gene indicated. The colours indicate the tissue of origin for each cell line, as indicated in the legend. The boxes themselves are <a href="https://en.wikipedia.org/wiki/Box_plot#Types_of_box_plots" title="Box plot">Tukey box plots</a> â€“ the bold line indicates the median of each group, the edges of the box indicates the interquartile range, and the whiskers indicate 1.5 times the interquartile range.</p>


<a name="gene_altered"><h2>When is a driver gene considered 'altered'?</h2></a>

<p> We integrate copy number profiles and exome sequencing to identify likely functional alterations in cancer driver genes. For most oncogenes we consider a functional alteration to be either an amplification, recurrent missense mutation, or recurrent in frame deletion or insertion. Recurrence is defined as as at least 3 previous mutations of a particular site in the <a href="http://cancer.sanger.ac.uk/cosmic">COSMIC database</a>. For a small number of oncogenes (ERBB2, MYC, MYCN) we consider only amplifications as functional events, while for another group (KRAS, BRAF, NRAS, HRAS) we only consider recurrent mutations/indels. In addition to recurrent missense or indel events, for tumour suppressors we consider that all nonsense, frameshift and splice-site mutations are likely functional alterations. We also consider that deletions (derived from copy number profiles) are functional alterations for tumour suppressors.</p>


<a name="genotyping"><h2>Where does the genotyping data come from?</h2></a>

<p> Exome data for ~1,000 cell lines are obtained from the <a href="http://www.cancerrxgene.org/gdsc1000/GDSC1000_WebResources/Home.html">GDSC resource</a>. Copy number data for the same
set of cell lines is obtained from <a href = "http://www.cancerrxgene.org/downloads">CancerRXGene</a>. We use the gene level copy number scores which are derived from PICNIC analysis of Affymetrix SNP6.0 array data. An oncogene is considered amplified if the entire coding sequence has 8 or more copies while a tumour suppressor is considered deleted if any part of the coding sequence has a copy number of 0. </p>

<a name="json"><h2>How can I access this data from scripts as a JSON web service?</h2></a>

<p>(1) This dependency data can also be obtained in JSON format for input into scripts. The data in the dependency table can be obtained, using a URL of the format:
<pre class="script">
http://{{current_url}}{% url 'gendep:get_dependencies'  search_by='driver' gene_name='GENE' histotype_name='TISSUE' study_pmid='STUDY' %}
</pre>
where:
<ul>
<li>"driver" to indicate search is by driver gene. Change this to "target" to search by target gene. </li>
<li>"GENE" is the name of the gene gene (eg. ERBB2), as listed on the <a href="{% url 'gendep:drivers' %}">drivers page</a> or <a href="{% url 'gendep:targets' %}">targets page</a>.</li>
<li>"TISSUE" is either: "ALL_HISTOTYPES", or one of the tissues (eg. LUNG) listed on the <a href="{% url 'gendep:tissues' %}">tissues page</a>.</li>
<li>"STUDY" is either "ALL_STUDIES", or one of the study PubMed Id's as listed on the <a href="{% url 'gendep:studies' %}">studies page</a>.</li>
</ul>

For example:
(a) To retreive data for driver gene ERBB2, all tissues and all studies:
<pre class="script">
http://{{current_url}}{% url 'gendep:get_dependencies' search_by='driver' gene_name='ERBB2' histotype_name='ALL_HISTOTYPES' study_pmid='ALL_STUDIES' %}
</pre>

(b) To retreive data for target gene ERBB2, Pan-cancer, study Campbell(2016):
<pre class="script">
http://{{current_url}}{% url 'gendep:get_dependencies' search_by='target' gene_name='ERBB2' histotype_name='PANCAN' study_pmid='26947069' %}
</pre>

<p>You can retrieve and extract the data from JSON, using a script. In Python 3, you could use:</p>
<pre class="script">
import urllib.request, json
url = "http://{{current_url}}{% url 'gendep:get_dependencies' search_by='driver' gene_name='ERBB2' histotype_name='BREAST' study_pmid='26947069' %}"
response = urllib.request.urlopen(url).read().decode('utf-8')
data = json.loads(response)
</pre>

<p>This data contains similar dependency information as the 'Download as CSV' link above the dependency table.</p>

<p>(2) The drivers can be obtained in JSON format using the URL.</p>
<pre class="script">
http://{{current_url}}{% url 'gendep:get_drivers' %}
</pre>

<p>Optionally a '?name=GENE_NAME' can be added to retreive only those drivers containing this string, eg:</p>
<pre class="script">
http://{{current_url}}{% url 'gendep:get_drivers' %}?name=ER
</pre>
which would retreive drivers containing 'ER', eg: ERBB2, RERE, DICER1, etc.


<p>(3) You can also retrieve information about one gene, by using:</p>
<pre class="script">
http://{{current_url}}{% url 'gendep:get_gene_info' gene_name='ERBB2' %}
</pre>

<p>(4) You can also retrieve the boxplot data in CSV (comma-separated-value) format using 'get_boxplot' followed by driver, target, tissue and study-pubmed-id:</p>
<pre class="script">
http://{{current_url}}{% url 'gendep:get_boxplot' dataformat='csv' driver_name='ERBB2' target_name='MAK2P' histotype_name='PANCAN' study_pmid='26947069' %}
</pre>

<a name="downloaddb"><h2>How can I download the whole CancerGD database to interrogate locally?</h2></a>
<p>The full CancerGD database can be downloaded as an <a href="{% static 'gendep/db_sqlite.zip' %}" download>Sqlite3 database</a> or a <a href="{% static 'gendep/all_dependencies.csv.gz' %}" download>Comma-separate values (CSV) file</a>. These are compressed files so need to be uncompressed using either Linux/MacOS command-line, eg: "unzip db_sqlite.zip" or "gunzip all_dependencies.csv.gz". On Windows could use a program such as <a href="http://www.7-zip.org/">7-Zip</a>.

<p>To query the SQLite database you can use a program such as the free <a href="http://sqlitebrowser.org/">SQLite Browser</a></p>
	
{% endblock %}
